segmentation/src/OpticalFlowSegmentation.cpp

```diff
@@ -0,0 +1,493 @@
+            WriteImage(
+                wholeChainPath,
+                draw_particle_on_slice_(currentCurve, zIndex, -1, true));
+        }
+
+        // Set up the maximum number of threads
+        auto maxThreads = std::thread::hardware_concurrency() - 1;
+        if (maxThreads_.has_value()) {
+            maxThreads = std::min(maxThreads, maxThreads_.value());
+        }
+
+        // Calculate the num threads we're going to use
+        const std::uint32_t minPointsPerThread{15};
+        const auto numPts = currentVs.size();
+        const auto ptsPerThread = static_cast<std::uint32_t>(std::floor(
+            static_cast<float>(numPts) /
+            static_cast<float>(minPointsPerThread)));
+        const auto numThreads =
+            std::max(1U, std::min(ptsPerThread, maxThreads));
+        const auto baseSegmentLength = static_cast<std::uint32_t>(std::floor(
+            static_cast<float>(numPts) / static_cast<float>(numThreads)));
+        const auto numThreadsWithExtraPoint = numPts % numThreads;
+
+        // Parallel computation of curve segments
+        std::vector<std::vector<Voxel>> subsegmentVectors;
+        std::size_t startIdx{0};
+        for (const auto& i : range(numThreads)) {
+            auto segmentLength =
+                baseSegmentLength + (i < numThreadsWithExtraPoint ? 1 : 0);
+            auto endIdx = startIdx + segmentLength;
+
+            // Change start_idx and end_idx to include overlap
+            auto startIdxPadded =
+                static_cast<std::int64_t>((i == 0) ? 0 : (startIdx - 2));
+            auto endIdxPadded = static_cast<std::int64_t>(
+                (i == numThreads - 1) ? numPts : (endIdx + 2));
+
+            // Copy from currentVs to our vector
+            auto startIt = std::next(currentVs.begin(), startIdxPadded);
+            auto endIt = std::next(currentVs.begin(), endIdxPadded);
+            subsegmentVectors.emplace_back(startIt, endIt);
+            startIdx = endIdx;
+        }
+
+        // Initialize the threads
+        std::vector<std::thread> threads;
+        std::vector<std::vector<Voxel>> subsegmentPoints(numThreads);
+        for (const auto& i : range(numThreads)) {
+            threads.emplace_back(
+                [this, &subsegmentVectors, &zIndex, &subsegmentPoints, i]() {
+                    const Chain subsegmentChain(subsegmentVectors[i]);
+                    const FittedCurve curve(subsegmentChain, zIndex);
+                    subsegmentPoints[i] = compute_curve_(curve, zIndex);
+                });
+        }
+
+        // Join threads and stitch curve segments together
+        for (auto& thread : threads) {
+            thread.join();
+        }
+
+        // Stitch curve segments together, discarding overlapping points
+        std::vector<Voxel> stitched;
+        stitched.reserve(currentVs.size());
+        for (auto [i, segment] : enumerate(subsegmentPoints)) {
+            auto startIt = segment.begin();
```

Comment:
Didn't seem like we actually needed to delete from `subsegmentPoints` but just copy a limited range to the stitched. Can you confirm this works as expected?